proxy/src/connection.rs

```diff
@@ -186,6 +189,61 @@ impl BoundPort {
     }
 }
 
+// ===== impl ConditionallyUpgradeServerToTls =====
+
+impl ConditionallyUpgradeServerToTls {
+    fn new(socket: TcpStream, tls: tls::ConnectionConfig<tls::ServerConfig>) -> Self {
+        ConditionallyUpgradeServerToTls::Plaintext(Some(ConditionallyUpgradeServerToTlsInner {
+            socket,
+            tls,
+            peek_buf: BytesMut::with_capacity(8192),
+        }))
+    }
+}
+
+impl Future for ConditionallyUpgradeServerToTls {
+    type Item = Connection;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            *self = match self {
+                ConditionallyUpgradeServerToTls::Plaintext(ref mut inner) => {
+                    let r = {
+                        let inner = inner.as_mut().unwrap();
+                        try_ready!(inner.socket.read_buf(&mut inner.peek_buf));
+                        tls::conditional_accept::match_client_hello(
+                            inner.peek_buf.as_ref(), &inner.tls.identity)
+                    };
+                    match r {
+                        tls::conditional_accept::Match::Matched => {
+                            trace!("upgrading accepted connection to TLS");
+                            let inner = inner.take().expect("Polled after ready");
+                            let upgrade_to_tls = tls::Connection::accept(
+                                inner.socket, inner.peek_buf.freeze(), inner.tls.config);
+                            ConditionallyUpgradeServerToTls::UpgradeToTls(upgrade_to_tls)
+                        },
+                        tls::conditional_accept::Match::NotMatched => {
+                            trace!("passing through accepted connection without TLS");
+                            let inner = inner.take().expect("Polled after ready");
+                            let conn = Connection::plain_with_peek_buf(
+                                inner.socket, inner.peek_buf, tls::ReasonForNoTls::NotProxyTls);
+                            return Ok(Async::Ready(conn));
+                        },
+                        tls::conditional_accept::Match::Incomplete => {
+                            return Ok(Async::NotReady);
```

Comment:
What will notify the task again if we return `NotReady` here?

Reply:
(I could be wrong, but I _think_ what we want to do in this case is to continue looping rather than returning `NotReady`?)

Reply:
You're right, @hawkw! Would you mind preparing a PR that addresses only this?

Reply:
Opened #1209. :)